demonstrations/tutorial_fdhs.py

```diff
@@ -0,0 +1,455 @@
+
+import numpy as np
+import pennylane as qml
+from pennylane import X, Y, Z
+
+import jax
+import jax.numpy as jnp
+import optax
+jax.config.update("jax_enable_x64", True)
+
+n_wires = 4
+gens = [X(i) @ X(i+1) for i in range(n_wires-1)]
+gens += [Y(i) @ Y(i+1) for i in range(n_wires-1)]
+gens += [Z(i) @ Z(i+1) for i in range(n_wires-1)]
+
+H = qml.sum(*gens)
+
+g = qml.lie_closure(gens)
+g = [op.pauli_rep for op in g]
+
+##############################################################################
+# 
+# Cartan decomposition
+# --------------------
+# 
+# A Cartan decomposition is a bipartition :math:`\mathfrak{g} = \mathcal{k} \oplus \mathcal{m}` into a vertical subspace
+# :math:`\mathfrak{k}` and an orthogonal horizontal subspace :math:`\mathfrak{m}`. In practice, it can be induced by an
+# involution function :math:`\Theta` that fulfils :math:`\Theta(\Theta(g)) = g \forall g \in \mathfrak{g}`. Different 
+# involutions lead to different types of Cartan decompositions, which have been fully classified by Cartan 
+# (see `wikipedia <https://en.wikipedia.org/wiki/Symmetric_space#Classification_result>`__).
+# 
+# .. note::
+#     Note that :math:`\mathfrak{k}` is the small letter k in
+#     `Fraktur <https://en.wikipedia.org/wiki/Fraktur>`__ and a 
+#     common - not our - choice for the vertical subspace in a Cartan decomposition.
+#
+# One common choice of involution is the so-called even-odd involution for Pauli words
+# :math:`P = P_1 \otimes P_2 .. \otimes P_n` where :math:`P_j \in \{I, X, Y, Z\}`.
+# It essentially counts whether the number of non-identity Pauli operators in the Pauli word is even or odd.
+
+def even_odd_involution(op):
+    """Generalization of EvenOdd to sums of Paulis"""
+    [pw] = op.pauli_rep
+    parity = len(pw) % 2
+
+    return parity
+
+even_odd_involution(X(0)), even_odd_involution(X(0) @ Y(3))
+
+##############################################################################
+# 
+# The vertical and horizontal subspaces are the two eigenspaces of the involution, corresponding to the :math:`\pm 1` eigenvalues.
+# In particular, we have :math:`\Theta(\mathfrak{k}) = \mathfrak{k}` and :math:`\Theta(\mathfrak{m}) = - \mathfrak{m}`.
+# So in order to perform the Cartan decomposition :math:`\mathfrak{g} = \mathfrak{k} \oplus \mathfrak{m}`, we simply
+# sort the operators by whether or not they yield a plus or minus sign from the involution function.
+
+def cartan_decomposition(g, involution):
+    """Cartan Decomposition g = k + m
+    
+    Args:
+        g (List[PauliSentence]): the (dynamical) Lie algebra to decompose
+        involution (callable): Involution function :math:`\Theta(\cdot)` to act on PauliSentence ops, should return ``0/1`` or ``True/False``.
+    
+    Returns:
+        k (List[PauliSentence]): the even parity subspace :math:`\Theta(\mathfrak{k}) = \mathfrak{k}`
+        m (List[PauliSentence]): the odd parity subspace :math:`\Theta(\mathfrak{m}) = \mathfrak{m}` """
```

Comment:
Horizontal scrollbar
```suggestion
        involution (callable): Involution function :math:`\Theta(\cdot)` to act on
            PauliSentence ops, should return ``0/1`` or ``True/False``.
    
    Returns:
        k (List[PauliSentence]): the even parity subspace :math:`\Theta(x) = x`
        m (List[PauliSentence]): the odd parity subspace :math:`\Theta(x) = -x` 
    """
```